prometheus_juju_exporter/collector.py

```diff
@@ -77,14 +77,18 @@ async def _connect_controller(
                 endpoint=endpoint, username=username, password=password, cacert=cacert
             )
 
-    async def _get_machines_in_model(self, uuid: str) -> Dict[Any, Any]:
+    async def _get_machines_in_model(self, uuid: str) -> Optional[Dict[Any, Any]]:
         """Get a list of all machines in the model with their stats.
 
         :return: status information for all machines in the model
         """
-        model = await self.controller.get_model(uuid)
-        status = await model.get_status()
-        await model.disconnect()
+        try:
+            model = await self.controller.get_model(uuid)
+            status = await model.get_status()
+            await model.disconnect()
+        except Exception as err:  # pylint: disable=W0703
```

Comment:
Can we use [JujuConnectionError](https://github.com/juju/python-libjuju/blob/master/juju/errors.py#L27)? It's better to catch exact error and otherwise raise an exception. For example if there is wrong username and password it should raise a proper exception instead of error log.

Reply:
I think it's OK in this case. We don't want an error with single model to stop whole exporter.

Reply:
Does not look good to me, but I can see your point.
What if specific [JujuConnectionError](https://github.com/juju/python-libjuju/blob/master/juju/errors.py#L27) will be catch with warning message and error message in other cases?

Reply:
Imo any exception caught here should be reported as "Error", not just "Warning", because it prevents proper data collection.

Reply:
The more I think about it, I think that we should catch proper errors here. These two `JujuError` and `SSLError` should cover all exceptions. Imagine the situation that libjuju will change and object `model` will no longer have `get_status()` function, in that case exporter should failed.

Reply:
I don't see any benefit in that approach. No one is going to notice dead exporter (errored unit). I think it's far better to make exporter resilient to errors and export them as metrics. Then we can have alerting rules tied to them and they can be properly caught by monitoring.

Reply:
This wide catch is intended because I see juju connection.connect has the possibility of [raising general Exception](https://github.com/juju/python-libjuju/blob/b90f16ba0d4670f159aaccda6dfab8135826abe4/juju/client/connection.py#L355) in some cases. 

And yes, this is a short term solution as described in the Jira task. The medium/long term solution is to have a separate metrics record any inaccessible models and alert on them. This solution requires COS integration and I already started looking into it